src/util/amount.rs

```diff
@@ -0,0 +1,549 @@
+                    }
+                }
+                b'.' => match decimal {
+                    None => decimal = Some(0),
+                    Some(_) => return Err(ParseAmountError::InvalidFormat),
+                },
+                _ => return Err(ParseAmountError::InvalidFormat),
+            }
+        }
+
+        let scaleup = match decimal {
+            Some(d) => max_decimals - d,
+            None => max_decimals,
+        };
+        for _ in 0..scaleup {
+            value = match 10i64.checked_mul(value) {
+                Some(v) => v,
+                None => return Err(ParseAmountError::TooBig),
+            };
+        }
+
+        if negative {
+            value *= -1;
+        }
+
+        Ok(Amount::from_inner(value))
+    }
+
+    /// Express this Amount as a floating-point value in the given denomination.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn float_denomination(&self, denom: Denomination) -> f64 {
+        let exp = denom.scale() - P::scale();
+        (self.0 as f64) * 10f64.powi(exp)
+    }
+
+    /// Format this Amount in the given denomination.
+    pub fn fmt_denomination(&self, f: &mut fmt::Write, denom: Denomination) -> fmt::Result {
+        let nb_decimals = (P::scale() - denom.scale()) as usize;
+        write!(f, "{:.*} {}", nb_decimals, self.float_denomination(denom), denom)
+    }
+
+    /// Get a formatter string of this Amount in the given denomination,
+    /// followed by the shorthand of the denomination.
+    pub fn to_denominated_string(&self, denom: Denomination) -> String {
+        let mut buf = String::new();
+        self.fmt_denomination(&mut buf, denom).unwrap();
+        buf
+    }
+
+    /// Get a new Amount by rescaling this one to a new precision.
+    /// Note that some decimals will get lost if the new precision is lower.
+    pub fn as_rescaled<O: Precision>(&self) -> Amount<O> {
+        let amt = rescale_inner(self.0, O::scale() - P::scale());
+        Amount::from_inner(amt)
+    }
+
+    /// Convert into a new Amount by rescaling this one to a new precision.
+    /// Note that some decimals will get lost if the new precision is lower.
+    pub fn into_rescaled<O: Precision>(self) -> Amount<O> {
+        let amt = rescale_inner(self.0, O::scale() - P::scale());
+        Amount::from_inner(amt)
+    }
+
+    /// The maximum value of an Amount.
+    pub fn max_value() -> Amount<P> {
+        Amount::from_inner(Inner::max_value())
+    }
+
+    /// The minimum value of an amount.
+    pub fn min_value() -> Amount<P> {
+        Amount::from_inner(Inner::min_value())
+    }
+}
+
+impl<P: Precision> fmt::Debug for Amount<P> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Amount({}, presision={})", self.0, P::scale())
+    }
+}
+
+impl fmt::Display for Amount {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.fmt_denomination(f, Denomination::Bitcoin)
```

Comment:
I think this doesn't work with round trips: it will output `x BTC` but `FromStr` expects `x`, doesn't it?

Reply:
ah damn, I thought fmt_denomination just wrote the value, not the suffix. you're right

Reply:
I updated the top post